jax/_src/cudnn/scaled_matmul_stablehlo.py

```diff
@@ -0,0 +1,730 @@
+
+
+Array = jnp.ndarray
+block_scaled_dot_name = "__op$block_scaled_dot"
+
+@dataclass
+class BlockScaleConfig:
+  mode: str
+  block_size: int
+  data_type: DTypeLike
+  scale_type: DTypeLike
+  global_scale: Array | None
+  infer_only: bool
+
+mxfp8_config = BlockScaleConfig(
+    mode='mxfp8',
+    block_size=32,
+    data_type=jnp.float8_e4m3fn,
+    scale_type=jnp.float8_e8m0fnu,
+    global_scale=None,
+    infer_only=False
+)
+
+BlockScaleConfigs = List[BlockScaleConfig]
+mxfp8_configs: BlockScaleConfigs = [mxfp8_config, mxfp8_config, mxfp8_config]
+
+def default_layouts(*shapes):
+  return [range(len(shape) - 1, -1, -1) for shape in shapes]
+
+def element_type_to_backend_config_type(dtype):
+  _element_type_to_backend_config_type_mapping = {
+      ir.BF16Type.get(): "BF16",
+      ir.F16Type.get(): "F16",
+      ir.F32Type.get(): "F32",
+  }
+  return _element_type_to_backend_config_type_mapping[dtype]
+
+
+def _scaled_matmul_impl(a, b, a_scale, b_scale, preferred_element_type):
+  return _scaled_matmul_p.bind(
+      a, b, a_scale, b_scale, preferred_element_type=preferred_element_type
+  )
+
+
+def _scaled_matmul_cuda_lowering(
+    ctx, a, b, a_scales, b_scales, preferred_element_type
+  ):
+  lhs_type = ir.RankedTensorType(a.type)
+  lhs_shape = lhs_type.shape
+  rhs_type = ir.RankedTensorType(b.type)
+  rhs_shape = rhs_type.shape
+
+  batch, non_contracting_lhs, contracting = lhs_shape
+  _, non_contracting_rhs, _ = rhs_shape
+  result_shape = (batch, non_contracting_lhs, non_contracting_rhs)
+
+  out_type = mlir.dtype_to_ir_type(preferred_element_type)
+  result_types = [ir.RankedTensorType.get(result_shape, out_type)]
+
+  operands = [a, b, a_scales, b_scales]
+  backend_config = {
+      "scaled_dot_backend_config": {
+          "lhs_batch_dimensions": [0],
+          "rhs_batch_dimensions": [0],
+          "dequantize_type": element_type_to_backend_config_type(out_type),
+      }
+  }
+
+  backend_config = json.dumps(backend_config)
+  out = mlir.custom_call(
+      block_scaled_dot_name,
+      result_types=result_types,
+      operands=operands,
+      backend_config=backend_config,
+      operand_layouts=default_layouts(
+          *[ir.RankedTensorType(operand.type).shape for operand in operands]
+      ),
+      result_layouts=default_layouts(result_shape),
+  )
+  return [out.result]
+
+
+def _scaled_matmul_abstract(a, b, a_scale, b_scale, *, preferred_element_type):
+  a_dtype = dtypes.canonicalize_dtype(a.dtype)
+  batch, non_contracting_lhs, contracting_lhs = a.shape
+  _, non_contracting_rhs, _ = b.shape
+  output_shape = (batch, non_contracting_lhs, non_contracting_rhs)
+  return (core.ShapedArray(output_shape, preferred_element_type),)
+
+
+_scaled_matmul_p = core.Primitive("scaled_matmul")
+_scaled_matmul_p.multiple_results = True
+_scaled_matmul_p.def_impl(partial(xla.apply_primitive, _scaled_matmul_p))
+_scaled_matmul_p.def_abstract_eval(_scaled_matmul_abstract)
+
+
+mlir.register_lowering(
```

Comment:
See [CI test failures](https://github.com/jax-ml/jax/actions/runs/13423839455/job/37506043305?pr=26345) - this code only runs on GPU now.

You will need to either: 
* add lowering fallback to XLA for other platforms like CPU and TPU, or
* explicitly say your implementation only works for certain platforms and prevent invalid usages: platform-check or require an explicit platform arg in the API (like this [cudnn implementation](https://github.com/jax-ml/jax/blob/main/jax/_src/nn/functions.py#L1019)), and disable tests on other platforms.